janitor/functions.py

```diff
@@ -3140,17 +3141,30 @@ def update_where(
         # 4 8  0
 
     :param df: The pandas DataFrame object.
-    :param conditions: conditions used to update a target column and target
-        value
+    :param conditions: Conditions used to update a target column
+        and target value.
     :param target_column_name: Column to be updated. If column does not exist
         in dataframe, a new column will be created; note that entries that do
         not get set in the new column will be null.
     :param target_val: Value to be updated
     :returns: An updated pandas DataFrame.
     :raises: IndexError if **conditions** does not have the same length as
         **df**.
+    :raises: TypeError if **conditions** is not a pandas-compatible string
+        query.
     """
-    df.loc[conditions, target_column_name] = target_val
+
+    # use query mode if a string expression is passed
+    if isinstance(conditions, str):
+        # get the index that meets the conditions criteria
+        conditions_index = df.query(conditions).index
+
+        # pass target_val to dataframe
+        df.loc[conditions_index, target_column_name] = target_val
+
+    else:
+        df.loc[conditions, target_column_name] = target_val
```

Comment:
I think there might be a cleaner DAG-like way of handling this.

Let me know whether this works, @samukweku.

```suggestion
    if isinstance(conditions, str):
        conditions_index = df.query(conditions).index
    else:
        conditions_index = df.loc[conditions].index   
    df.loc[conditions_index, target_column_name] = target_val
```

I think this makes the code a bit easier to maintain long-term, by setting up the index in the if/else block, and only finally at the end doing the assignment. There's a nice "parallel track" based on condition. With that, the in-line comments aren't really needed either.

Reply:
much cleaner code. and true, easier to maintain long term.

Reply:
If you don’t mind I would love to test this new code before you merge. Need to check that index is only for Boolean True

Reply:
Yes, absolutely, that's the right way to approach this! Thanks @samukweku!